### ISSUE TITLE
Should /var/lib/ubuntu-advantage/private/ be world-listable?

### ISSUE THREAD

**Issue by OddBloke** [CanonicalLtd/ubuntu-advantage-client#645]
The _files_ in `/var/lib/ubuntu-advantage/private/` are only readable by root, but any user on the system can list the files that it contains.

(Currently, all this allows them to determine is a list of entitlements which they could glean from `ua status` anyway.)

**Comment by powersj** [comment]
@panlinux thoughts?

**Comment by panlinux** [comment]
The way it is now, it doesn't leak information not available elsewhere. But we are relying on some assumptions:
a) that even in the future getting directory contents is no big deal
b) that whatever is creating those files never forgets to make them 0600 root:root and isn't subjective to races (create && change perms && add content)
It would be another layer of security to have the directory 700 root:root

**Comment by powersj** [comment]
I think this is worth revisiting during 19.7

**Comment by panlinux** [comment]
The packaging bit is simple, but that only takes care of upgrades. Since that directory is created when the machine is attached for the first time, that means the python code should take care of it, or else the directory will remain with perms `0755` until the next package upgrade.
The only place I found within the code that does it, I think, is `makedirs()` in `uaclient/config.py:write_cache()`:
```python
    def write_cache(self, key: str, content: "Any") -> None:
        filepath = self.data_path(key)
        data_dir = os.path.dirname(filepath)
        if not os.path.exists(data_dir):
            os.makedirs(data_dir)
...
```
Is that the place?
Alternatively, I could preemptively create the private dir in `postinst`:
```diff
--- a/debian/postinst
+++ b/debian/postinst
@@ -72,6 +72,18 @@ case "$1" in
       fi
       chmod 0600 /var/log/ubuntu-advantage.log
       chown root:root /var/log/ubuntu-advantage.log
+      data_dir="/var/lib/ubuntu-advantage"
+      private_dir="$data_dir/private"
+      if [ ! -d "$data_dir" ]; then
+          mkdir -m 0755 -p "$data_dir"
+      else
+          chmod 0755 "$data_dir"
+      fi
+      if [ ! -d "$private_dir" ]; then
+          mkdir -m 0700 -p "$private_dir"
+      else
+          chmod 0700 "$private_dir"
+      fi
       ;;
 esac
```
I could also make this a bit more robust by getting the value of `data_dir` from `/etc/ubuntu-advantage/uaclient.conf`.

**Comment by blackboxsw** [comment]
Fixed by #793
